pkg/index/index.go

```diff
@@ -0,0 +1,206 @@
+	Copy() Node
+}
+
+type node struct {
+	Node
+	parent   *node
+	children []*node
+	mask     uint32
+}
+
+func (n *node) addChild(child *node) {
+	n.children = append(n.children, child)
+}
+
+type Index struct {
+	root        *node
+	lookupTable map[string]*node
+	l           sync.Mutex
+}
+
+func NewIndex() *Index {
+	i := &Index{
+		lookupTable: make(map[string]*node),
+	}
+
+	return i
+}
+
+// Insert inserts a copy of the given node to the tree under the given parent.
+func (i *Index) Insert(n Node) error {
+	i.l.Lock()
+	defer i.l.Unlock()
+
+	_, ok := i.lookupTable[n.String()]
+	if ok {
+		return fmt.Errorf("node %s already exists in index", n.String())
+	}
+
+	log.Debugf("Inserting %s (parent: %s) in index", n.String(), n.Parent())
+
+	newNode := &node{
+		Node: n.Copy(),
+	}
+
+	if n.Parent() == n.String() {
+		if i.root != nil {
+			return fmt.Errorf("node cannot point to self unless it's root")
+		}
+
+		// set root
+		i.root = newNode
+
+	} else {
+		p, ok := i.lookupTable[n.Parent()]
+		if !ok {
+			return fmt.Errorf("Can't find parent %s", n.Parent())
+		}
+		newNode.parent = p
+		p.addChild(newNode)
+	}
+
+	i.lookupTable[n.String()] = newNode
+	return nil
+}
+
+// Get returns a Copy of the named node.
+func (i *Index) Get(nodeId string) (Node, error) {
+	i.l.Lock()
+	defer i.l.Unlock()
+
+	n, ok := i.lookupTable[nodeId]
+	if !ok {
+		return nil, ErrNodeNotFound
+	}
+
+	return n.Copy(), nil
+}
+
+func (i *Index) List() ([]Node, error) {
+	i.l.Lock()
+	defer i.l.Unlock()
+
+	nodes := make([]Node, 0, len(i.lookupTable))
+
+	for _, v := range i.lookupTable {
+		nodes = append(nodes, v.Copy())
+	}
+
+	return nodes, nil
+}
+
+// Delete deletes a leaf node
+func (i *Index) Delete(nodeId string) (Node, error) {
+	i.l.Lock()
+	defer i.l.Unlock()
+
+	n, ok := i.lookupTable[nodeId]
+	if !ok {
+		return nil, fmt.Errorf("Node %s not found", nodeId)
+	}
+
+	if len(n.children) != 0 {
+		return nil, fmt.Errorf("Node %s has children %#q", nodeId, n.children)
+	}
+
+	// remove the reference to the node from its parent
+	parent := n.parent
+	var deleted bool
+	for idx, child := range parent.children {
+		if child.String() == nodeId {
+			parent.children = append(parent.children[:idx], parent.children[idx+1:]...)
+			deleted = true
+		}
+	}
+
+	if !deleted {
+		err := fmt.Errorf("%s not found in tree", nodeId)
+		log.Errorf("%s", err)
+		return nil, err
+	}
+
+	// remove from the lookup table
+	delete(i.lookupTable, nodeId)
+	n.parent = nil
+
+	return n.Node, nil
+}
+
+type iterflag int
+
+const (
+	NOOP iterflag = iota
+	STOP
+)
+
+type visitor func(Node) (iterflag, error)
+
+func (i *Index) bfs(root *node, visitFunc visitor) error {
+	i.l.Lock()
+	defer i.l.Unlock()
+
+	// XXX Look into parallelizing this without breaking API boundaries.
+	return i.bfsworker(root, func(n *node) (iterflag, error) { return visitFunc(n.Node) })
+}
+
+func (i *Index) bfsworker(root *node, visitFunc func(*node) (iterflag, error)) error {
+
+	queue := list.New()
```

Comment:
I tried using a `List` here, and this definitely works, but the extra pointer (because `list.Element` has a `Prev`,`Next`, and Value pointer) is an extra 16bytes per element.  This vs the 1 reference (8byte) for each element in the array.  I'd rather take the realloc overhead of copying vs keeping more references in memory.